runtime_scan/pkg/orchestrator/scanconfigwatcher/schedule.go

```diff
@@ -0,0 +1,264 @@
+		return t, true
+	}
+
+	return from, false
+}
+
+func isCronPeriodic(c *cronexpr.Expression) bool {
+	// NOTE: `from.Add(1)` is needed as `from` represents zero time and cronexpr returns empty list if it is provided
+	//       with zero time as first parameter. Non-standard cron expressions (Quartz) may include year field
+	//       representing time range which might be in the past hence the `from` time is set to zero time.
+	from := time.Date(1, 1, 1, 0, 0, 0, 0, time.UTC)
+	var numOfCalculatedTimes uint = 2
+
+	return len(c.NextN(from.Add(1), numOfCalculatedTimes)) == int(numOfCalculatedTimes)
+}
+
+type ScheduleState int8
+
+const (
+	ScheduleStateDisabled ScheduleState = iota
+	ScheduleStateUnscheduled
+	ScheduleStateNotDue
+	ScheduleStateDue
+	ScheduleStateOverdue
+)
+
+func (s ScheduleState) String() string {
+	switch s {
+	case ScheduleStateDisabled:
+		return "Disabled"
+	case ScheduleStateUnscheduled:
+		return "Unscheduled"
+	case ScheduleStateNotDue:
+		return "NotDue"
+	case ScheduleStateDue:
+		return "Due"
+	case ScheduleStateOverdue:
+		return "OverDue"
+	default:
+		return "Unknown"
+	}
+}
+
+type ScanConfigSchedule struct {
+	State         ScheduleState
+	OperationTime *OperationTime
+	Window        *ScheduleWindow
+}
+
+func (w ScanConfigSchedule) String() string {
+	return fmt.Sprintf("state: %s, operation time: [%s], schedule window: [%s]", w.State, w.OperationTime, w.Window)
+}
+
+// nolint:cyclop
+func NewScanConfigSchedule(scanConfig *models.ScanConfig, window *ScheduleWindow) (*ScanConfigSchedule, error) {
+	if scanConfig.Disabled != nil && *scanConfig.Disabled {
+		return &ScanConfigSchedule{
+			State:  ScheduleStateDisabled,
+			Window: window,
+		}, nil
+	}
+
+	if scanConfig.Scheduled == nil || (scanConfig.Scheduled.CronLine == nil && scanConfig.Scheduled.OperationTime == nil) {
+		return &ScanConfigSchedule{
+			State:  ScheduleStateUnscheduled,
+			Window: window,
+		}, nil
+	}
+
+	var cronExpr *cronexpr.Expression
+	var err error
+	if scanConfig.Scheduled.CronLine != nil {
+		cronExpr, err = cronexpr.Parse(*scanConfig.Scheduled.CronLine)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse cron expression %s: %w", *scanConfig.Scheduled.CronLine, err)
+		}
+	}
+
+	var oTime time.Time
+	if scanConfig.Scheduled.OperationTime != nil {
+		oTime = (*scanConfig.Scheduled.OperationTime).UTC()
+	}
+
+	operationTime := NewOperationTime(oTime, cronExpr)
+
+	if window.In(operationTime.Time()) {
+		return &ScanConfigSchedule{
+			State:         ScheduleStateDue,
+			OperationTime: operationTime,
+			Window:        window,
+		}, nil
+	}
+
+	if window.Before(operationTime.Time()) {
+		if !operationTime.IsRecurring() {
+			return &ScanConfigSchedule{
+				State:         ScheduleStateUnscheduled,
+				OperationTime: operationTime,
+				Window:        window,
+			}, nil
+		}
+		return &ScanConfigSchedule{
+			State:         ScheduleStateOverdue,
+			OperationTime: operationTime,
+			Window:        window,
+		}, nil
+	}
```

Comment:
Can you add a comment here? The `window.Before(operationTime.Time())` is a bit confusing for me, does it means that the operation time is before the window or after the window?
As I understand it, the time is before the window, so not sure why it is `Overdue`?

Reply:
I missed the part that `ScheduleStateOverdue` will be handled later and will update the operation time.
Thanks for the clarification.